docs/developer-docs/smart-contracts/best-practices/idempotency.mdx

```diff
@@ -0,0 +1,90 @@
+---
+keywords: [beginner, concept, best practices, developer]
+---
+
+import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
+
+# Retries and Idempotency
+
+In the case of network or other issues, ICP clients (such as agents) that issue
+ingress update calls may become unable to determine whether their ingress
+request has been processed. For example, this can happen if the client loses its
+connection until the request status has been removed from the state tree (recall
+that ICP will remove the request from the system state tree some time after
+ingress expiry).
+
+Thus, it is important to design and/or use canister APIs such that it is possible to retry requests safely, even when the ICP doesn't provide any information as to previous request attempts. This page describes general approaches that both the canister authors and the clients can adopt to enable safe retries.
+
+## Idempotent Canister APIs
+
+We say that a canister endpoint is idempotent if it can be called multiple times without changing the result or canister state beyond the first call. Whenever an endpoint is idempotent or can be made idempotent by the developer, this provides an easy way to implement safe retries.
```

Comment:
```suggestion
An endpoint of a canister is considered idempotent if calling it multiple times does not alter the result or the state of the canister after the initial call. When an endpoint is idempotent or can be made idempotent by the developer, it offers a straightforward approach to implementing safe retries.
```

Reply:
Is "without changing the result or canister state beyond the first call" accurate? It seems that it (1) assumes that the first call actually reaches the canister (otherwise a retry after an initial call that failed because it didn't even reach the canister would be required to fail as well, which is not the intention of idempotency, right?), and (2) ignores canister state that will change on every call like cycles balance.

Maybe this is fine for the introduction, but wanted to raise it nevertheless to see what you think.

Reply:
It could say "beyond the first executed call".

And leaving out "changing the state" and focusing only on the result. Although I guess that simply producing the same result does not imply idempotency; so maybe "without meaningfully altering the state".

Reply:
Is it incorrect if we leave out "meaningfully"? Isn't any state change violating idempotency?

Reply:
I guess David's point is that the canister state will change on any call, whether it's the cycle balance, incidentally changing the size of the heap or bumping a counter of "total requests handled".

Idempotency in this context is more about only applying the change _relevant to the specific call_ exactly once. So e.g. I would consider an account transfer idempotent as long as the transfer happened exactly once, even if some small fee was deducted from my account every time I retried said call. In that context I care primarily about making a payment first and foremost and not so much about how much I'm charged for it and when (assuming it's a reasonably small fee).

Reply:
I shortened this to "executing the endpoint multiple times is equivalent to executing it once" and added a footnote to comment on the fact that we take the "user intent" into the account when we say "equivalent".

Reply:
I'll resolve this to be able to merge, hopefully everyone is happy with the current wording - shout if not.